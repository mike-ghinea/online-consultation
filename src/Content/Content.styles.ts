import styled from "styled-components";

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  margin-top: auto;
`;

const QuestionWrapper = styled.div<{ $isVisible: boolean }>`
  display: none;
  ${(props) =>
    props.$isVisible &&
    `
    display: flex;
    flex-direction: column;
    `};
`;

const Title = styled.h1`
  font-size: 2rem;
  @media (min-width: 480px) {
    font-size: 4rem;
  }
`;

const SubTitle = styled.h2`
  font-size: 1rem;
  @media (min-width: 480px) {
    font-size: 2rem;
  }
`;

const ButtonWrapper = styled.div`
  display: flex;
  justify-content: center;
  gap: 2rem;
  padding-bottom: 2rem;
`;

const Button = styled.button`
  font-size: 1rem;
  min-width: 3rem;
  border-radius: 12px;
  @media (min-width: 480px) {
    font-size: 3rem;
    min-width: 6rem;
  }

  background-color: rgb(0, 63, 130);
  color: white;
  &:hover {
    cursor: pointer;
  }
  &:hover,
  &:focus {
    box-shadow: 0 0 0 5px #24cb34;
  }
`;

const ThankYouWrapper = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
`;

export default {
  ContentWrapper,
  QuestionWrapper,
  Title,
  SubTitle,
  ButtonWrapper,
  Button,
  ThankYouWrapper,
};
