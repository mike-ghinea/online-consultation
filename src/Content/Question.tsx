import s from "./Content.styles";

const Question: React.FC<{
  question: { text: string; key: string };
  isVisible: boolean;
  incrementVisibleQuestion: () => void;
  updateAnswers: (key: string, value: boolean) => void;
}> = ({ question, isVisible, incrementVisibleQuestion, updateAnswers }) => {
  return (
    <s.QuestionWrapper $isVisible={isVisible}>
      <s.Title>{question.text}</s.Title>
      <s.ButtonWrapper onClick={incrementVisibleQuestion}>
        <s.Button
          onClick={() => updateAnswers(question.key, true)}
          data-testid={`t_${question.key}_yes`}
        >
          Yes
        </s.Button>
        <s.Button
          onClick={() => updateAnswers(question.key, false)}
          data-testid={`t_${question.key}_no`}
        >
          No
        </s.Button>
      </s.ButtonWrapper>
    </s.QuestionWrapper>
  );
};

export default Question;
