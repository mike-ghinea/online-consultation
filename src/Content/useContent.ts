import { useEffect, useState } from "react";
import { ConsultationAnswers } from "../requests/types";
import { PostConsultationAnswers } from "../requests/requests";

export const useContent = () => {
  const [visibleQuestion, setVisibleQuestion] = useState(0);
  const [answers, setAnswers] = useState<ConsultationAnswers>({
    valid_age: false,
    likes_fruit: false,
    previous_allergy: false,
    allergy_medication: false,
    other_medication: false,
  });

  const questions = [
    {
      text: "Are you aged between 0 and 100?",
      key: "valid_age",
    },
    {
      text: "Do you like eating fruit?",
      key: "likes_fruit",
    },
    {
      text: "Have you ever had an allergic reaction to the Genovian pear?",
      key: "previous_allergy",
    },
    {
      text: "Are you currently taking any allergy medication?",
      key: "allergy_medication",
    },
    {
      text: "Are you currently taking any other medication?",
      key: "other_medication",
    },
  ];

  useEffect(() => {
    if (visibleQuestion > questions.length - 1)
      PostConsultationAnswers(answers);
  }, [answers, visibleQuestion]);

  const incrementVisibleQuestion = () => {
    setVisibleQuestion((prev) => prev + 1);
  };

  const updateAnswers = (key: string, value: boolean) => {
    setAnswers((prev) => {
      return {
        ...prev,
        [key]: value,
      };
    });
  };

  return {
    updateAnswers,
    incrementVisibleQuestion,
    visibleQuestion,
    questions,
  };
};
