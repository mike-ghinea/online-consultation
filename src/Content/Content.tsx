import s from "./Content.styles";
import Question from "./Question";
import { useContent } from "./useContent";

const Content: React.FC = () => {
  const {
    questions,
    visibleQuestion,
    incrementVisibleQuestion,
    updateAnswers,
  } = useContent();

  const questionList = questions.map((question, index) => (
    <Question
      key={question.key}
      question={question}
      isVisible={visibleQuestion === index}
      incrementVisibleQuestion={incrementVisibleQuestion}
      updateAnswers={updateAnswers}
    />
  ));

  return (
    <s.ContentWrapper role="main">
      {visibleQuestion > 4 ? (
        <s.ThankYouWrapper>
          <s.Title>Thank you!</s.Title>
          <s.SubTitle>
            One of our pharmacists will get in touch with you soon.
          </s.SubTitle>
        </s.ThankYouWrapper>
      ) : (
        <>{questionList}</>
      )}
    </s.ContentWrapper>
  );
};
export default Content;
