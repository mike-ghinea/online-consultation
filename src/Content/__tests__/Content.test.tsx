import Content from "../Content";
import { fireEvent, render } from "@testing-library/react";
import Question from "../Question";

describe("Content", () => {
  it("should render correctly", () => {
    const { baseElement } = render(<Content />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should show the thank you page", () => {
    const { getByTestId, getByText } = render(<Content />);
    fireEvent.click(getByTestId("t_valid_age_yes"));
    fireEvent.click(getByTestId("t_likes_fruit_yes"));
    fireEvent.click(getByTestId("t_previous_allergy_yes"));
    fireEvent.click(getByTestId("t_allergy_medication_yes"));
    fireEvent.click(getByTestId("t_other_medication_yes"));
    expect(getByText("Thank you!")).toBeTruthy();
  });
});

describe("Question", () => {
  it("should call the answer updater correctly", () => {
    const updateAnswersSpy = jest.fn();
    const { getByTestId } = render(
      <Question
        question={{
          text: "test",
          key: "key",
        }}
        isVisible={false}
        incrementVisibleQuestion={jest.fn()}
        updateAnswers={updateAnswersSpy}
      />,
    );

    fireEvent.click(getByTestId("t_key_yes"));

    expect(updateAnswersSpy).toHaveBeenCalledWith("key", true);

    fireEvent.click(getByTestId("t_key_no"));

    expect(updateAnswersSpy).toHaveBeenCalledWith("key", false);
  });
});
