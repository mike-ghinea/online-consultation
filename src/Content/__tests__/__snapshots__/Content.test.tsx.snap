// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content should render correctly 1`] = `
<body>
  <div>
    <div
      class="sc-blHHSb gLAEXT"
      role="main"
    >
      <div
        class="sc-gtLWhw bhtvLb"
      >
        <h1
          class="sc-egkSDF iafcA-D"
        >
          Are you aged between 0 and 100?
        </h1>
        <div
          class="sc-dntaoT cqssgI"
        >
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_valid_age_yes"
          >
            Yes
          </button>
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_valid_age_no"
          >
            No
          </button>
        </div>
      </div>
      <div
        class="sc-gtLWhw crRhgt"
      >
        <h1
          class="sc-egkSDF iafcA-D"
        >
          Do you like eating fruit?
        </h1>
        <div
          class="sc-dntaoT cqssgI"
        >
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_likes_fruit_yes"
          >
            Yes
          </button>
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_likes_fruit_no"
          >
            No
          </button>
        </div>
      </div>
      <div
        class="sc-gtLWhw crRhgt"
      >
        <h1
          class="sc-egkSDF iafcA-D"
        >
          Have you ever had an allergic reaction to the Genovian pear?
        </h1>
        <div
          class="sc-dntaoT cqssgI"
        >
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_previous_allergy_yes"
          >
            Yes
          </button>
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_previous_allergy_no"
          >
            No
          </button>
        </div>
      </div>
      <div
        class="sc-gtLWhw crRhgt"
      >
        <h1
          class="sc-egkSDF iafcA-D"
        >
          Are you currently taking any allergy medication?
        </h1>
        <div
          class="sc-dntaoT cqssgI"
        >
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_allergy_medication_yes"
          >
            Yes
          </button>
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_allergy_medication_no"
          >
            No
          </button>
        </div>
      </div>
      <div
        class="sc-gtLWhw crRhgt"
      >
        <h1
          class="sc-egkSDF iafcA-D"
        >
          Are you currently taking any other medication?
        </h1>
        <div
          class="sc-dntaoT cqssgI"
        >
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_other_medication_yes"
          >
            Yes
          </button>
          <button
            class="sc-ivxoEo eQSpqy"
            data-testid="t_other_medication_no"
          >
            No
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
